### PR TITLE
nanoplot

### PR DESCRIPTION
**Pull request by sumanm99**
## Issue ticket number

Closes #xxxx 

## Checklist before requesting a review

- [x] I have performed a self-review of my code

- [x] Conforms to the [Contributing guidelines](https://github.com/viash-hub/base/blob/main/CONTRIBUTING.md)

- [ ] Proposed changes are described in the [CHANGELOG.md](https://github.com/viash-hub/base/blob/main/CHANGELOG.md)

- [ ] I have tested my code with `viash ns test --parallel -q <name or namespace>`

- Check the correct box. Does this PR contain:
  - [ ] Breaking changes
  - [x] New functionality
  - [ ] Major changes
  - [ ] Minor changes
  - [ ] Documentation
  - [ ] Bug fixes

### CODE REVIEW

File: src/nanoplot/config.vsh.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    path: test_data
+engines:
+  - type: docker
+    image: quay.io/biocontainers/nanoplot:1.42.0--pyhdfd78af_0
+    setup:
+      - type: docker
+        run: |
+          touch ./var/software_versions.txt && \
+          version=$(docker run quay.io/biocontainers/nanoplot:1.42.0--pyhdfd78af_0 NanoPlot --version) && \
+          echo "$version" > ./var/software_versions.txt
```
Comment (rcannood) [inline src/nanoplot/config.vsh.yaml:61]: * This code is run inside the docker. You're running docker in docker with this command.
* Make sure that the format of the resulting file is valid yaml, i.e. the contents of the file should be `NanoPlot: "1.42.0"`.
* Output the file to `/var/software_versions.txt`, not `/var/software_versions.txt`

Check the other components for valid examples

File: src/nanoplot/test.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+echo "> Run NanoPlot"
+"$meta_executable" \
+  --fastq "$meta_resources_dir/test_data/real.fastq" \
+  -o "output"
+
+echo ">> Check if output directory is empty"
+if [ -z "$(ls -A ./output)" ]; then
+    echo "No output"
+    exit 1
+fi
```
Comment (rcannood) [inline src/nanoplot/test.sh:19]: Check whether the contents of the resulting file is correct

File: var/software_versions.txt
Comment (rcannood) [inline var/software_versions.txt:1]: Delete this file

**Review by rcannood — CHANGES_REQUESTED** [review]

File: src/nanoplot/config.vsh.yaml (hunk 1 above)
Comment (sumanm99) [reply src/nanoplot/config.vsh.yaml:61]: I think I have output the file to the correct destination?

File: .gitignore [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 # IDE ignores
 .idea/
+.vscode/
```
Comment (rcannood) [inline .gitignore:6]: Please undo the changes in this file

File: src/nanoplot/test.sh [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+## VIASH START
+meta_executable="$PWD/target/executable/nanoplot/nanoplot"
+meta_resources_dir="$PWD/src/nanoplot"
+## VIASH END
+
+###########################################################################
+
+# Test 1: Run NanoPlot with only input parameter
+
+echo "> Run Test 1: one input"
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:14]:
```suggestion
echo "> Run Test 1: one input"

mkdir "$TMPDIR/test1" && pushd "$TMPDIR/test1" > /dev/null
```

File: src/nanoplot/test.sh [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+fi
+
+rm -f output/* # Clear output
+
+echo "Test 1 succeeded."
+
+###########################################################################
+
+# Test 2: Run NanoPlot with multiple inputs
+
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:51]:
```suggestion
mkdir "$TMPDIR/test2" && pushd "$TMPDIR/test2" > /dev/null
```

File: src/nanoplot/test.sh [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+fi
+
+rm -f output/* # Clear output
+
+echo "Test 2 succeeded."
+
+###########################################################################
+
+# Test 3: Run NanoPlot with multiple options-1
+
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:89]:
```suggestion
mkdir "$TMPDIR/test3" && pushd "$TMPDIR/test3" > /dev/null
```

File: src/nanoplot/test.sh [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+  echo "Extracted data is empty."
+  exit 1
+fi
+
+# Check if the output file starts with "biobox" prefix
+if ! ls output/biobox* > /dev/null 2>&1; then
+    echo "The prefix is not added to the output files."
+fi
+
+rm -f output/* # Clear output
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:141]:
```suggestion
popd > /dev/null
```

File: src/nanoplot/test.sh [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+echo "Test 3 succeeded."
+
+###########################################################################
+
+# Test 4: Run NanoPlot with multiple options-2
+
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:148]:
```suggestion
mkdir "$TMPDIR/test4" && pushd "$TMPDIR/test4" > /dev/null
```

File: src/nanoplot/test.sh [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+fi
+
+rm -f output/* # Clear output
+
+echo "Test 4 succeeded."
+
+###########################################################################
+
+# Test 5: Run NanoPlot with different input
+
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:195]:
```suggestion
mkdir "$TMPDIR/test5" && pushd "$TMPDIR/test5" > /dev/null
```

File: src/nanoplot/test.sh [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+fi
+
+rm -f output/* # Clear output
+
+echo "Test 5 succeeded."
+
+###########################################################################
+
+# Test 6: Run NanoPlot with output options
+
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:233]:
```suggestion
mkdir "$TMPDIR/test6" && pushd "$TMPDIR/test6" > /dev/null
```

File: src/nanoplot/test.sh [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+if find out -name "*.png" -type f -size 0 | grep -q .; then 
+  echo "At least one plot is empty."
+  exit 1
+fi
+if find out -name "*.txt" -type f -size 0 | grep -q .; then 
+  echo "NanoPlot summary file is empty."
+  exit 1
+fi
+
+rm -rf out/ # Clear output
```
Comment (jakubmajercik) [inline src/nanoplot/test.sh:265]:
```suggestion
popd > /dev/null
```

File: src/nanoplot/test_data/test.bam
Comment (jakubmajercik) [inline src/nanoplot/test_data/test.bam:1]: this file is too large, is it possible to get a smaller one that fits under the 100KB limit?

File: src/nanoplot/test_data/test.pickle
Comment (jakubmajercik) [inline src/nanoplot/test_data/test.pickle:1]: is this file being used?

File: src/nanoplot/test_data/test_rich.fastq
Comment (jakubmajercik) [inline src/nanoplot/test_data/test_rich.fastq:1]: is this file being used?

File: src/nanoplot/test_data/test_summary.txt
Comment (jakubmajercik) [inline src/nanoplot/test_data/test_summary.txt:1]: is this file being used?

**Review by jakubmajercik — CHANGES_REQUESTED** [review]
A few suggestions regarding the config and script but mainly about testing.
WDYT @rcannood ?

File: src/nanoplot/config.vsh.yaml [hunk 12]
```diff
@@ -0,0 +1,241 @@
+name: nanoplot
+description: |
+  Run NanoPlot on nanopore-sequenced reads.
+keywords: ["quality control", "qc", "fastq", "sequencing summary", "nanopore"]
```
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/config.vsh.yaml:4]: Ok. 
I just copied the same keywords from the nf-core nanoplot yaml file [https://github.com/nf-core/modules/blob/master/modules/nf-core/nanoplot/meta.yml](url).

File: src/nanoplot/test.sh
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test.sh:1]: One file of each type is sufficient to run the tool.

[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test.sh:1]: Ok. I'll add test cases for each of the file format.

File: src/nanoplot/config.vsh.yaml [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        multiple: true
+        multiple_sep: " "
+  - name: Outputs
+    arguments:
+      - name: --outdir
+        alternatives: [-o]
+        type: file
+        direction: output
+        description: Specify directory in which output has to be created.
+        default: output
```
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/config.vsh.yaml:99]: But the user can choose to not specify the argument too, in which case the output files will go to the default directory, right?

File: src/nanoplot/test_data/test.fasta.gz
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test_data/test.fasta.gz:1]: I haven't used it in the test script, but I have tested all the file formats which is why they are there. I have found a smaller fasta file that works.

File: src/nanoplot/test_data/test_ubam.bam
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test_data/test_ubam.bam:1]: All other test files I have are within the 100KB limit, except for the ubam file. I could not find another example of unmapped bam. I'm not sure how we can identify them as well as it's extension is .bam itself.

I guess I will exclude ubam files just like we did for cram files.

File: src/nanoplot/test_data/test_ubam.bam.bai
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test_data/test_ubam.bam.bai:1]: I have removed ubam files.

File: src/nanoplot/test.sh [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+## VIASH START
+meta_executable="$PWD/target/executable/nanoplot/nanoplot"
+meta_resources_dir="$PWD/src/nanoplot"
+## VIASH END
+
+###########################################################################
+
+# Test 1: Run NanoPlot with only input parameter
+
```
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test.sh:13]: But what is the point of creating temporary directories when by default the output will be saved in the default output directory in the config file?

File: .gitignore [hunk 15]
```diff
@@ -14,4 +14,4 @@ target/
 
 # nextflow specific ignores
 .nextflow*
-work
```
Comment (rcannood) [inline .gitignore:17]: Undo this change

File: src/nanoplot/config.vsh.yaml [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+argument_groups:
+  - name: Inputs
+    arguments:
+      - name: --fastq
+        type: file
+        description: Input fastq file(s).
+        example: read.fq
+        direction: input
+        multiple: true
+        multiple_sep: " "
```
Comment (rcannood) [inline src/nanoplot/config.vsh.yaml:22]: Don't override the `multiple_sep`, leave it as the default `;`.

File: src/nanoplot/config.vsh.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (rcannood) [reply src/nanoplot/config.vsh.yaml:100]: Indeed -- please remove the default value and set the `--outdir` as `required: true`

File: src/nanoplot/test.sh
[the remark this answers is not in the document]
Comment (rcannood) [reply src/nanoplot/test.sh:1]: As mentioned by Jakub, I think that the test data is a bit too big (See [Contributing guidelines](https://github.com/viash-hub/biobox/blob/main/CONTRIBUTING.md#step-6-create-or-fetch-test-data)). Can you generate smaller versions of the .gz, pickle, and feather files such that they are all <1KB?

File: src/nanoplot/test.sh (hunk 14 above)
[the remark this answers is not in the document]
Comment (rcannood) [reply src/nanoplot/test.sh:13]: I don't think you need to create a separate TMPDIR per se, since the unit test will already be executed in a temp dir.

However I do agree that it would be nice if each separate test is run in its own directory, to make sure that you don't get any data leakage.

Example:

```bash
mkdir test1
pushd test1 # cd into directory
echo "> Run Test 1: one input (Fastq)"
# ... test code ...
popd # go back to original directory

mkdir test2
pushd test2
echo "> Run Test 2: multiple inputs (Fastq)"
# ... test code ...
popd
```

**Review by rcannood — CHANGES_REQUESTED** [review]
Hey @sumanm99 ! Great work on this component! Everything looks mostly ok to me, except for a few remaining minor issues.

Could you resolve the last few issues so we can merge approve this PR?

File: src/nanoplot/test.sh
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test.sh:1]: I have maintained all the file size to be less than 100 KB, as is the maximum limit in the Contributing Guidelines.

Using smaller size files (with say less than 5 reads and inadequate data) does not generate the desired output with the plots and Summary statistics. That's why I have tried to use better files with less than 100KB limit.

File: .gitignore (hunk 15 above)
Comment (sumanm99) [reply .gitignore:17]: I see that work is included in the .gitignore file. I haven't made any changes to the original .gitignore file of biobox. Could this be an old commit?

Comment (rcannood) [reply .gitignore:17]: No, you removed a newline at the end of this file

File: src/nanoplot/test.sh
[the remark this answers is not in the document]
Comment (rcannood) [reply src/nanoplot/test.sh:1]: I will reiterate. I think the test data is too big. I know the contributing guidelines state that the test data can be <100KB if need be, but I don't think such large data is necessary in this case. Please check if you can generate the .gz, .pickle, and .feather file at runtime or reduce them such that they are <1KB.

**Review by rcannood — DISMISSED** [review]

File: src/nanoplot/test.sh
[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test.sh:1]: I cannot get these files on runtime because I had obtained them from a GitHub repo, and using git clone in the test script results in an error "git command not found" in viash_automount.
So I will rather remove those test files.

[the remark this answers is not in the document]
Comment (sumanm99) [reply src/nanoplot/test.sh:1]: I have generated the said files at runtime in the test script.

**Review by emmarousseau — APPROVED** [review]
Changes made by myself

**Review by jakubmajercik — APPROVED** [review]
